lib/utils/logger.js

```diff
@@ -246,6 +249,84 @@ const logger = {
 		clearTimeout(delayedLog);
 		console.log(jsonRenderer.render(obj));
 	},
+
+	/**
+	 * @typedef {['object', object]
+	 * | ['array', any[]]
+	 * | ['element', ElementStruct]
+	 * | ['function', string]
+	 * | ['table', [any[], any[]]]
+	 * | ['trace', StackTrace]} ComplexStruct
+	 *
+	 * @typedef {'log' | 'error' | 'warn' | 'info' | 'debug' | 'assert' | 'clear' | 'count' | 'countReset' |
+	 * 'dir' | 'dirxml' | 'group' | 'groupCollapsed' | 'groupEnd' | 'table' | 'time' | 'timeEnd' |
+	 * 'timeLog' | 'trace'} Subtype
+	 */
+
+	/**
+	 * Log console output from device which are delivered in WS 'console' messages.
+	 * @param {Subtype} subtype - console method
+	 * @param {string | Array<string | number | boolean | null | ComplexStruct} [data] - data to log
+	 * @param {number} [timestamp] - required for subtype 'time' | 'timeEnd' | 'timeLog'
+	 */
+	appOutput: (subtype, data, timestamp) => {
+		if (config.logLevel === logLevels.silent) {
+			return;
+		}
+
+		// Time logs and trace are displayed with console.log and data from device.
+		const method = ['time', 'timeEnd', 'timeLog', 'trace'].includes(subtype) ? 'log' : subtype;
+
+		if (typeof data === 'string' || typeof data === 'undefined') {
+			data = [data];
+		}
+
+		// Timestamps are delivered from device.
+		if (['time', 'timeEnd', 'timeLog'].includes(subtype)) {
+			const timer = data[0] || 'default';
+
+			if (subtype === 'time') {
+				logger.consoleTimestamps[timer] = timestamp;
+			} else {
+				if (logger.consoleTimestamps[timer]) {
+					console.log(`${timer}: ${timestamp - logger.consoleTimestamps[timer]}ms`);
+				} else {
+					console.warn(`Timer '${timer}' does not exist`);
+				}
+				if (subtype === 'timeEnd') {
+					delete logger.consoleTimestamps[timer];
+				}
+			}
+		} else {
+			data.forEach(d => {
+				if (!Array.isArray(d)) {
+					// If not an array display as it is.
+					console[method](data);
+				} else if (d[0] === 'element') {
+					// If element, diaplay as DOM el with textContent if present, do not show inner elements.
+					if (d[1].nodeType === 3) {
+						console[method](d[1].nodeValue);
+					} else {
+						const attrs = d[1].attributes
+							? Object.entries(d[1].attributes).map(([attr, val]) => ` ${attr}="${val}"`).join('')
+							: '';
+						const child = d[1].children && d[1].children[0] && d[1].children[0].nodeType === 3
```

Comment:
In case of `<div>some text<span>more text</span></div>` only 'some text' is displayed and nothing shows that there is something else? and probably `d[1].children[0]` check is redundant as `children` prop presents only if there is at least one child

Reply:
We did not want to go very deep and decided to only display text if text node is there.
Open for suggestions.

Reply:
Maybe we can add ... at the end if text node is to big or there is some other elements

Reply:
Chrome has limit of 80 chars, if element has more text it just displays ...